core/node/node_framework/src/implementations/layers/reorg_detector.rs

```diff
@@ -0,0 +1,93 @@
+use std::sync::Arc;
+
+use anyhow::Context;
+use zksync_block_reverter::BlockReverter;
+use zksync_core::reorg_detector::{self, ReorgDetector};
+
+use crate::{
+    implementations::resources::{
+        main_node_client::MainNodeClientResource,
+        pools::{MasterPool, PoolResource},
+        reverter::BlockReverterResource,
+    },
+    precondition::Precondition,
+    service::{ServiceContext, StopReceiver},
+    task::Task,
+    wiring_layer::{WiringError, WiringLayer},
+};
+
+#[derive(Debug)]
+pub struct ReorgDetectorLayer;
+
+#[async_trait::async_trait]
+impl WiringLayer for ReorgDetectorLayer {
+    fn layer_name(&self) -> &'static str {
+        "reorg_detector_layer"
+    }
+
+    async fn wire(self: Box<Self>, mut context: ServiceContext<'_>) -> Result<(), WiringError> {
+        // Get resources.
+        let main_node_client = context.get_resource::<MainNodeClientResource>().await?.0;
+
+        let pool_resource = context.get_resource::<PoolResource<MasterPool>>().await?;
+        let pool = pool_resource.get().await?;
+
+        let reverter = context.get_resource::<BlockReverterResource>().await?.0;
+
+        // Create and insert precondition.
+        context.add_precondition(Box::new(ReorgDetectorPrecondition {
+            reorg_detector: ReorgDetector::new(main_node_client.clone(), pool.clone()),
+            reverter,
+        }));
+
+        // Create and insert task.
+        context.add_task(Box::new(ReorgDetectorTask {
+            reorg_detector: ReorgDetector::new(main_node_client, pool),
+        }));
+
+        Ok(())
+    }
+}
+
+pub struct ReorgDetectorPrecondition {
+    reorg_detector: ReorgDetector,
+    reverter: Arc<BlockReverter>,
+}
+
+#[async_trait::async_trait]
+impl Precondition for ReorgDetectorPrecondition {
+    fn name(&self) -> &'static str {
+        "reorg_detector"
+    }
+
+    async fn check(mut self: Box<Self>, _stop_receiver: StopReceiver) -> anyhow::Result<()> {
+        match self.reorg_detector.check_consistency().await {
+            Ok(()) => {}
+            Err(reorg_detector::Error::ReorgDetected(last_correct_l1_batch)) => {
+                tracing::info!("Reverting to l1 batch number {last_correct_l1_batch}");
+                self.reverter.roll_back(last_correct_l1_batch).await?;
+                tracing::info!("Revert successfully completed");
+            }
+            Err(err) => return Err(err).context("reorg_detector.check_consistency()"),
+        }
+        Ok(())
+    }
+}
```

Comment:
I don't think it would work in general case, e.g. for EN running in a distributed mode (e.g. once we will be able to run EN on multiple machines, like we do with the main node). This precondition must be on every machine, since we don't want the components to start if there was a reorg, but at the same time if reorg will be detected, _every machine_ will attempt to do a rollback.

Instead, probably we need to change it as follows:

1. Precondition would simply check if there exists a reorg in a loop. If there is, the task would just sleep for a second. If not, the loop exits. This way every machine with this precondition would simply not start until the reorg is fixed.
2. You will also have an unconstrained oneshot task that will also check whether there is a reorg, and will perform a rollback if so. This task should be added by a separate layer, given that it will only run on one of the components, unlike the reorg detector.

Does it make sense?

Reply:
>Does it make sense?

1st - yes, 2nd - not really, to be more precise...this part
*given that it will only run on one of the components, unlike the reorg detector.*
particularly. Does it mean that in the distributed mode, only a single instance of EN has to revert?

Reply:
By distributed mode I mean that, say, API runs on one machine and state keeper + tree run on another one.

When you revert blocks, you make changes in Postgres and RocksDB, so you don't want two different machines to try to do it in parallel (moreover, API machines won't have RocksDB at all), so it has to be one machine that actually performs the rollback.